### PR TITLE
Set the filter for nullable unique indexes using a convention

### LINKED ISSUE #6794: Model Definition: Ability to override IS NOT NULL clauses of an index
> This is for SQL Server and I am using EF Core 1.0.1. 

Assume that I have the below entity:

``` csharp
public class Person
{
    public string Name { get; set; }
    public string Lastname { get; set; }
    public DateTime? DeletedOn { get; set; }
}
```

Inside `OnModelCreating`, I also have this to create a multi-column unique index:

``` csharp
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<Person>().HasIndex(x => new { x.Name, x.Lastname, x.DeletedOn }).IsUnique();
}
```

This causes EF Core to generate something similar to below:

``` csharp
CREATE UNIQUE NONCLUSTERED INDEX [IX_People_Name_Lastname_DeletedOn] ON [dbo].[People]
(
    [Name] ASC,
    [Lastname] ASC,
    [DeletedOn] ASC
)
WHERE ([Name] IS NOT NULL AND [Lastname] IS NOT NULL AND [DeletedOn] IS NOT NULL)
```

However, I want not to allow `DeletedOn` column to be null for this index. So, I am try to make EF Core to generate the below index creation script for me:

``` csharp
CREATE UNIQUE NONCLUSTERED INDEX [IX_People_Name_Lastname_DeletedOn] ON [dbo].[People]
(
    [Name] ASC,
    [Lastname] ASC,
    [DeletedOn] ASC
)
WHERE ([Name] IS NOT NULL AND [Lastname] IS NOT NULL)
```

**QUESTION: Is this possible to do this with EF Core fluent API?**

I dug a bit into the fluent API and was able to find some things related to `HasAnnotation` API but was not able to understand enough to achieve this.

### LINKED ISSUE #6901: SQL Server: Set nullable unique index filter using convention

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #6794

### CODE REVIEW

**Comment by AndriySvyryd** [comment]
@bricelam or @smitpatel could you take a look?

File: src/Microsoft.EntityFrameworkCore.Specification.Tests/TestUtilities/AnnotationComparer.cs [hunk 1]
```diff
@@ -30,7 +31,8 @@ public bool Equals(IAnnotation x, IAnnotation y)
             }
 
             return x.Name == y.Name
-                   && x.Value.Equals(y.Value);
+                   && (x.Name == CoreAnnotationNames.ValueGeneratorFactoryAnnotation
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/TestUtilities/AnnotationComparer.cs]: Why is this needed for the change?

Comment (AndriySvyryd) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/TestUtilities/AnnotationComparer.cs]: To allow the ModelBuilder tests that now run with SqlServer conventions (that add a discriminator) to pass

**Comment by smitpatel** [comment]
:shipit:

File: src/Microsoft.EntityFrameworkCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs [hunk 2]
```diff
@@ -421,25 +421,6 @@ protected override void Generate(DropTableOperation operation, IModel model, Mig
...
-                        if (i != 0)
-                        {
-                            builder.Append(" AND ");
-                        }
-
-                        builder
-                            .Append(SqlGenerationHelper.DelimitIdentifier(nullableColumns[i]))
-                            .Append(" IS NOT NULL");
-                    }
-                }
```
Comment (bricelam) [inline src/Microsoft.EntityFrameworkCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs]: Where does it get translated to SQL now?

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs]: It will be stored on `index.SqlServer().Filter` by conventions wherever applicable. Migrations use it from there.
